src/sentry/unmerge.py

```diff
@@ -0,0 +1,215 @@
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def stop_snuba_replacement(self, eventstream_state: Any) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def run_postgres_replacement(
+        self, project: Project, destination_id: int, locked_primary_hashes: Collection[str]
+    ) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_activity_args(self) -> Mapping[str, Any]:
+        raise NotImplementedError()
+
+
+@dataclass(frozen=True)
+class PrimaryHashUnmergeReplacement(UnmergeReplacement):
+    """
+    The "classical unmerge": Moving events out of the group based on primary_hash.
+    """
+
+    fingerprints: Collection[str]
+
+    def get_unmerge_key(
+        self, event: Event, locked_primary_hashes: Collection[str]
+    ) -> Optional[str]:
+        primary_hash = event.get_primary_hash()
+        if primary_hash in self.fingerprints and primary_hash in locked_primary_hashes:
+            return _DEFAULT_UNMERGE_KEY
+
+        return None
+
+    @property
+    def primary_hashes_to_lock(self) -> Collection[str]:
+        return self.fingerprints
+
+    def start_snuba_replacement(self, project: Project, source_id: int, destination_id: int) -> Any:
+        return eventstream.start_unmerge(project.id, self.fingerprints, source_id, destination_id)
+
+    def stop_snuba_replacement(self, eventstream_state: Any) -> None:
+        eventstream.end_unmerge(eventstream_state)
+
+    def run_postgres_replacement(
+        self, project: Project, destination_id: int, locked_primary_hashes: Collection[str]
+    ) -> None:
+        # Move the group hashes to the destination.
+        GroupHash.objects.filter(project_id=project.id, hash__in=locked_primary_hashes).update(
+            group=destination_id
+        )
+
+    def get_activity_args(self) -> Mapping[str, Any]:
+        return {"fingerprints": self.fingerprints}
+
+
+@dataclass(frozen=True)
+class UnmergeArgsBase:
+    """
+    Parsed arguments of the Sentry unmerge task. Since events of the source
+    issue are processed in batches, one can think of each batch as belonging to
+    a state in a statemachine.
+
+    That statemachine has only two states: Processing the first page
+    (`InitialUnmergeArgs`), processing second, third, ... page
+    (`SuccessiveUnmergeArgs`). On the first page postgres hashes are migrated,
+    activity models are created, eventstream and pagination state is
+    initialized, and so the successive tasks need to carry significantly more
+    state with them.
+    """
+
+    project_id: int
+    source_id: int
+    replacement: UnmergeReplacement
+    actor_id: Optional[int]
+    batch_size: int
+
+    @staticmethod
+    def parse_arguments(
+        project_id: int,
+        source_id: int,
+        destination_id: Optional[int],
+        fingerprints: Sequence[str],
+        actor_id: Optional[int],
+        last_event: Optional[str] = None,
```

Comment:
Is it definitely a string? In your tests I think you have this as `{"hello": "world"}`

Reply:
you are absolutely right